File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/converter/AddableOverlay.kt
package dev.note11.flutter_naver_map.flutter_naver_map.converter

import android.content.Context
import com.naver.maps.map.overlay.LocationOverlay
import com.naver.maps.map.overlay.Overlay
import dev.note11.flutter_naver_map.flutter_naver_map.controller.NaverMapControlHandler
import dev.note11.flutter_naver_map.flutter_naver_map.controller.overlay.OverlayHandler
import dev.note11.flutter_naver_map.flutter_naver_map.model.enum.NOverlayType
import dev.note11.flutter_naver_map.flutter_naver_map.model.flutter_default_custom.NPoint
import dev.note11.flutter_naver_map.flutter_naver_map.model.flutter_default_custom.NSize
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.info.NOverlayInfo
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay.NArrowheadPathOverlay
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay.NCircleOverlay
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay.NGroundOverlay
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay.NInfoWindow
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay.NMarker
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay.NMultipartPathOverlay
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay.NPathOverlay
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay.NPolygonOverlay
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay.NPolylineOverlay
import dev.note11.flutter_naver_map.flutter_naver_map.util.DisplayUtil

internal abstract class AddableOverlay<T : Overlay> {
    abstract val info: NOverlayInfo
    abstract fun createMapOverlay(): T

    private lateinit var commonProperties: Map<String, Any>

    fun setCommonProperties(rawArgs: Map<String, Any>) {
        commonProperties = rawArgs.filter { OverlayHandler.allPropertyNames.contains(it.key) }
    }

    fun applyCommonProperties(applyProperty: (name: String, arg: Any) -> Unit) {
        commonProperties.forEach { applyProperty(it.key, it.value) }
    }

    companion object {
        /** Used on @see [NaverMapControlHandler.addOverlayAll] */
        fun fromMessageable(
            info: NOverlayInfo,
            args: Map<String, Any>,
            context: Context,
        ): AddableOverlay<out Overlay> {
            val creator = when (info.type) {
                NOverlayType.MARKER -> NMarker::fromMessageable
                NOverlayType.INFO_WINDOW -> { rawMap ->
                    NInfoWindow.fromMessageable(rawMap, context = context)
                }
                NOverlayType.CIRCLE_OVERLAY -> NCircleOverlay::fromMessageable
                NOverlayType.GROUND_OVERLAY -> NGroundOverlay::fromMessageable
                NOverlayType.POLYGON_OVERLAY -> NPolygonOverlay::fromMessageable
                NOverlayType.POLYLINE_OVERLAY -> NPolylineOverlay::fromMessageable
                NOverlayType.PATH_OVERLAY -> NPathOverlay::fromMessageable
                NOverlayType.MULTIPART_PATH_OVERLAY -> NMultipartPathOverlay::fromMessageable
                NOverlayType.ARROWHEAD_PATH_OVERLAY -> NArrowheadPathOverlay::fromMessageable
                NOverlayType.LOCATION_OVERLAY -> throw IllegalArgumentException("LocationOverlay can not be created from json")
            }

            return creator.invoke(args).apply { setCommonProperties(args) }
        }

        fun LocationOverlay.toMessageable(): Map<String, Any?> = mapOf(
            "info" to NOverlayInfo.locationOverlayInfo.toMessageable(),
            anchorName to NPoint.fromPointF(anchor).toMessageable(),
            circleColorName to circleColor,
            circleOutlineColorName to circleOutlineColor,
            circleOutlineWidthName to DisplayUtil.pxToDp(circleOutlineWidth),
            circleRadiusName to DisplayUtil.pxToDp(circleRadius),
            iconSizeName to NSize.fromPixelSize(iconWidth, iconHeight).toMessageable(),
            subAnchorName to NPoint.fromPointF(subAnchor).toMessageable(),
            subIconSizeName to NSize.fromPixelSize(subIconWidth, subIconHeight)
                .toMessageable(),
        ) + overlayToMessageable(this)

        private const val anchorName = "anchor"
        private const val circleColorName = "circleColor"
        private const val circleOutlineColorName = "circleOutlineColor"
        private const val circleOutlineWidthName = "circleOutlineWidth"
        private const val circleRadiusName = "circleRadius"
        private const val iconSizeName = "iconSize"
        private const val subAnchorName = "subAnchor"
        private const val subIconSizeName = "subIconSize"
    }
}

private fun overlayToMessageable(overlay: Overlay): Map<String, Any?> = overlay.run {
    mapOf(
        OverlayHandler.zIndexName to zIndex,
        OverlayHandler.globalZIndexName to globalZIndex,
        OverlayHandler.isVisibleName to isVisible,
        OverlayHandler.minZoomName to minZoom,
        OverlayHandler.maxZoomName to maxZoom,
        OverlayHandler.isMinZoomInclusiveName to isMinZoomInclusive,
        OverlayHandler.isMaxZoomInclusiveName to isMaxZoomInclusive,
    )
}

File: ios/flutter_naver_map/Sources/flutter_naver_map/model/map/overlay/overlay/AddableOverlay.swift
import NMapsMap

internal protocol AddableOverlay: LazyOrAddableOverlay {
    override associatedtype OverlayType: NMFOverlay = NMFOverlay
    
    var info: NOverlayInfo { get }
    
    var overlayPayload: Dictionary<String, Any?> { get set }
    
    override func createMapOverlay() -> OverlayType
    
    func applyAtRawOverlay(_ overlay: OverlayType) -> OverlayType
    
    static func fromMessageable(_ v: Any) -> Self
}

func asAddableOverlayFromMessageable(info: NOverlayInfo, json: Any) throws -> any AddableOverlay {
    let d = asDict(json)
    let creator: (Any) -> any AddableOverlay
    switch info.type {
    case .marker: creator = NMarker.fromMessageable
    case .infoWindow: creator = NInfoWindow.fromMessageable
    case .circleOverlay: creator = NCircleOverlay.fromMessageable
    case .groundOverlay: creator = NGroundOverlay.fromMessageable
    case .polygonOverlay: creator = NPolygonOverlay.fromMessageable
    case .polylineOverlay: creator = NPolylineOverlay.fromMessageable
    case .pathOverlay: creator = NPathOverlay.fromMessageable
    case .multipartPathOverlay: creator = NMultipartPathOverlay.fromMessageable
    case .arrowheadPathOverlay: creator = NArrowheadPathOverlay.fromMessageable
    case .locationOverlay: throw NSError(domain: "asAddableOverlay", code: 0, userInfo: [NSLocalizedDescriptionKey: "LocationOverlay can not be created from json"])
    case .clusterableMarker: throw NSError(domain: "asAddableOverlay", code: 0, userInfo: [NSLocalizedDescriptionKey: "ClusterableMarker is not addableOverlay"])
    }
    return asAddableOverlayFromMessageableCorrector(json: d, creator: creator)
}

func asAddableOverlayFromMessageableCorrector(
    json: Dictionary<String, Any>,
    creator: (Any) -> any AddableOverlay
) -> any AddableOverlay {
    var overlay = creator(json)
    overlay.setCommonProperties(rawArgs: json)
    return overlay
}

internal extension AddableOverlay {
    mutating func setCommonProperties(rawArgs: Dictionary<String, Any?>) {
        overlayPayload = rawArgs.filter { allPropertyNames.contains($0.key) }
    }
    
    func applyCommonProperties(applyProperty: (_ name: String, _ arg: Any?) -> Void) {
        overlayPayload.forEach { applyProperty($0, $1) }
    }
}

extension NMFLocationOverlay {
    func toMessageable() -> Dictionary<String, Any?> {
        [
            "info": NOverlayInfo.locationOverlayInfo.toMessageable(),
            Self.anchorName: NPoint.fromCGPointWithOutDisplay(anchor).toMessageable(),
            Self.circleColorName: circleColor.toInt(),
            Self.circleOutlineColorName: circleOutlineColor.toInt(),
            Self.circleOutlineWidthName: circleOutlineWidth,
            Self.circleRadiusName: circleRadius,
            Self.iconSizeName: NSize(width: iconWidth, height: iconHeight).toMessageable(),
            Self.positionName: location.toMessageable(),
            Self.subAnchorName: NPoint.fromCGPointWithOutDisplay(subAnchor).toMessageable(),
            Self.subIconSizeName: NSize(width: subIconWidth, height: subIconHeight).toMessageable(),
        ].merging(overlayToMessageable(self), uniquingKeysWith: { $1 })
    }
    
    private static let anchorName = "anchor"
    private static let circleColorName = "circleColor"
    private static let circleOutlineColorName = "circleOutlineColor"
    private static let circleOutlineWidthName = "circleOutlineWidth"
    private static let circleRadiusName = "circleRadius"
    private static let iconSizeName = "iconSize"
    private static let positionName = "position"
    private static let subAnchorName = "subAnchor"
    private static let subIconSizeName = "subIconSize"
}

private func overlayToMessageable(_ overlay: NMFOverlay) -> Dictionary<String, Any?> {
    [
        zIndexName: overlay.zIndex,
        globalZIndexName: overlay.globalZIndex,
        isVisibleName: !overlay.hidden,
        minZoomName: overlay.minZoom,
        maxZoomName: overlay.maxZoom,
        isMinZoomInclusiveName: overlay.isMinZoomInclusive,
        isMaxZoomInclusiveName: overlay.isMaxZoomInclusive,
    ]
}
